tools/pythonapi/dftbplus.py

```diff
@@ -0,0 +1,253 @@
+        # and 2darray in restypes and argtypes specifications
+        self._dp1d = ndpointer(dtype=np.float64, ndim=1, flags='C_CONTIGUOUS')
+        self._dp2d = ndpointer(dtype=np.float64, ndim=2, flags='C_CONTIGUOUS')
+
+        # Register supported (DFTB+) routines
+        self._setup_interface()
+
+        if logfile is not None:
+            output_str = ctypes.create_string_buffer(str.encode(logfile))
+            self._dftbpluslib.dftbp_init(self._dftb_handler, output_str)
+        else:
+            self._dftbpluslib.dftbp_init(self._dftb_handler, None)
+
+        self._dftbpluslib.dftbp_get_input_from_file(self._dftb_handler,
+                                                    input_str, self._dftb_input)
+
+        self._dftbpluslib.dftbp_process_input(self._dftb_handler,
+                                              self._dftb_input)
+
+        self._natoms = self._dftbpluslib.dftbp_get_nr_atoms(self._dftb_handler)
+
+
+    def set_geometry(self, coords, latvecs=None):
+        '''Sets up the desired geometry.
+
+        Args:
+            coords (2darray):    absolute atomic positions
+                                 (in atomic units)
+            latvecs (2darray):   lattice vectors (in atomic units)
+                                 (None for non-periodic structures)
+        '''
+
+        periodic = latvecs is not None
+
+        if periodic:
+            latvecs = np.array(latvecs, dtype=float)
+            self._dftbpluslib.dftbp_set_coords_and_lattice_vecs(
+                self._dftb_handler, coords, latvecs)
+        else:
+            self._dftbpluslib.dftbp_set_coords(self._dftb_handler, coords)
+
+
+    def set_external_potential(self, extpot, extpotgrad=None):
+        '''Sets up an external potential.
+
+        Args:
+            extpot (2darray):     External potential at the position of each
+                                  atom. Shape: [natom]. (in atomic units)
+            extpotgrad (2darray): Gradient of the external potential at each
+                                  atom. Shape: [natom, 3]. (in atomic units)
+                                  This parameter is optional, you can pass None
+                                  if you did not ask DFTB+ to calculate forces.
+        '''
+
+        if extpotgrad is not None:
+            extpotgrad = np.ctypeslib.as_ctypes(extpotgrad)
+
+        self._dftbpluslib.dftbp_set_external_potential(
+            self._dftb_handler, extpot, extpotgrad)
+
+
+    def get_nr_atoms(self):
+        '''Queries the number of atoms.
+
+        Returns:
+            self._natoms (int): number of atoms
+        '''
+
+        return self._natoms
+
+
+    def get_energy(self):
+        '''Performs the energy (Mermin free) calculation
+           and queries the energy of the current geometry.
+
+        Returns:
+            energy[0] (float): calculated Mermin free energy
+                               (in atomic units)
+        '''
+
+        energy = np.zeros(1)
```

Comment:
`np.empty()`?

Reply:
Applied in 57a8873581acd0755186f6aeb363fd3feca0aab3